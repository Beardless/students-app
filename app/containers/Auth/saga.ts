/**
 * Gets the repositories of the user from Github
 */

import { call, put, select, takeEvery } from "redux-saga/effects";
import { AUTH_ACTION, requestUrl, userLoginInterface } from "./constants";

// @ts-ignore
import request from "utils/request";
import { authActionError, authActionSuccess } from "./actions";
import {
  getAuthorizationHeaders,
  getToken,
  setToken
} from "../../utils/userUtils";
import { makeSelectAuthError } from "./selectors";

/**
 * Github repos request/response handler
 */
export function* getTokenAsync(payload: userLoginInterface) {
  try {
    const options = {
      method: "POST",
      body: JSON.stringify(payload),
      headers: getAuthorizationHeaders()
    };
    const res = yield call(request, `${requestUrl}/login_check`, options);

    setToken(res.token);
  } catch (err) {
    yield put(authActionError(err.message));
  }
}

export function* getUser() {
  // TODO: Add func to get user
}

export function* login({ payload }) {
  try {
    yield call(getTokenAsync, payload);

    const err = yield select(makeSelectAuthError());

    if (!err) {
      const user = { username: "Bartek" };

      yield put(authActionSuccess(user));
    }
  } catch (err) {
    yield put(authActionError(err.message));
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export default function* authFlow() {
  // Watches for LOAD_REPOS actions and calls getRepos when one comes in.
  // By using `takeLatest` only the result of the latest API call is applied.
  // It returns task descriptor (just like fork) so we can continue execution
  // It will be cancelled automatically on component unmount
  // @ts-ignore
  yield takeEvery(AUTH_ACTION, login);
}
