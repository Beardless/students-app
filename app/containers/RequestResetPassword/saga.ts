import { call, put, takeEvery } from "redux-saga/effects";
import request from "../../utils/request";
import {
  requestResetPasswordActionError,
  requestResetPasswordActionSuccess
} from "./actions";
import { REQUEST_RESET_PASSWORD } from "./constants";
import { getAuthorizationHeaders } from "../../utils/userUtils";

function* requestResetPassword({ payload }: { payload: string }) {
  try {
    const options = {
      method: "POST",
      url: `/users/password/reset_request`,
      data: JSON.stringify({ email: payload }),
      headers: getAuthorizationHeaders()
    };

    console.log(options);

    yield call(request, options);

    yield put(requestResetPasswordActionSuccess(payload));
  } catch (err) {
    const { message } = err.response.data;
    yield put(requestResetPasswordActionError(message));
  }
}

export default function* requestResetPasswordFlow() {
  // @ts-ignore
  yield takeEvery(REQUEST_RESET_PASSWORD, requestResetPassword);
}
