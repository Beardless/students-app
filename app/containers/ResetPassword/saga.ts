import { call, put, takeEvery } from "redux-saga/effects";
import { RESET_PASSWORD_ACTION } from "./constatns";
import { getAuthorizationHeaders } from "../../utils/userUtils";
import request from "../../utils/request";
import {
  resetPasswordActionError,
  resetPasswordActionSuccess
} from "./actions";

export function* resetPassword({
  payload: { urlToken, ...rest }
}: {
  payload: { password: string; repeatedPassword: string; urlToken: string };
}) {
  try {
    const options = {
      method: "POST",
      url: `/users/password/reset?token=${urlToken}`,
      data: JSON.stringify({ ...rest }),
      headers: getAuthorizationHeaders()
    };

    yield call(request, options);

    yield put(resetPasswordActionSuccess());
  } catch (err) {
    const { message } = err.response.data;
    yield put(resetPasswordActionError(message));
  }
}

export default function* resetPasswordFlow() {
  // @ts-ignore
  yield takeEvery(RESET_PASSWORD_ACTION, resetPassword);
}
