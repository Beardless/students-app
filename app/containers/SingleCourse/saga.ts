import { COURSE_ACTION } from "./constants";
import { call, put, takeEvery } from "redux-saga/effects";
import { getAuthorizationHeaders } from "../../utils/userUtils";
import { courseActionError, courseActionSuccess } from "./actions";

// @ts-ignore
import request from "utils/request";

export function* getCourse({ payload }) {
  try {
    const { id: courseId } = payload;
    const options = {
      method: "GET",
      url: `/courses/${courseId}/modules`,
      headers: getAuthorizationHeaders()
    };
    const { data: modules } = yield call(request, options);

    yield put(courseActionSuccess(modules));
  } catch (err) {
    const { message } = err.response.data;
    yield put(courseActionError(message));
  }
}

export default function* getSingleCourse() {
  // @ts-ignore
  yield takeEvery(COURSE_ACTION, getCourse);
}
