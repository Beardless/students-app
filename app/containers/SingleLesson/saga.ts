import { LESSON_ACTION, LESSON_COMPLETE_ACTION } from "./constants";
import { call, put, takeEvery } from "redux-saga/effects";
import { getAuthorizationHeaders } from "../../utils/userUtils";
// @ts-ignore
import {
  lessonActionError,
  lessonActionSuccess,
  lessonCompleteActionError,
  lessonCompleteActionSuccess
} from "./actions";

// @ts-ignore
import request from "utils/request";
import { LessonInterface } from "./interfaces";

export function* getLesson({
  payload: { id: lessonId }
}: {
  payload: { id: string };
}) {
  try {
    const options = {
      method: "GET",
      url: `/lessons/${lessonId}`,
      headers: getAuthorizationHeaders()
    };
    const { data: lesson }: { data: LessonInterface } = yield call(
      request,
      options
    );

    yield put(lessonActionSuccess(lesson));
  } catch (err) {
    const { message } = err.response.data;
    yield put(lessonActionError(message));
  }
}

export function* completeLesson({
  payload: { isComplete, lessonId }
}: {
  payload: { isComplete: boolean; lessonId: string };
}) {
  try {
    const options = {
      method: "PATCH",
      url: `/lessons/${lessonId}/progress`,
      data: JSON.stringify({
        completed: isComplete
      }),
      headers: getAuthorizationHeaders()
    };

    const { data: lesson }: { data: LessonInterface } = yield call(
      request,
      options
    );

    yield put(lessonCompleteActionSuccess(lesson));
  } catch (err) {
    const { message } = err.response.data;
    yield put(lessonCompleteActionError(message));
  }
}

export default function* getSingleLesson() {
  // @ts-ignore
  yield takeEvery(LESSON_ACTION, getLesson);
  // @ts-ignore
  yield takeEvery(LESSON_COMPLETE_ACTION, completeLesson);
}
