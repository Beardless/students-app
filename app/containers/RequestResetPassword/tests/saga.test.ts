import { requestResetPassword } from "../saga";
import { put } from "redux-saga/effects";
import {
  requestResetPasswordActionError,
  requestResetPasswordActionSuccess
} from "../actions";

const requestResetPasswordPayload = { payload: "randomString" };

describe("requestResetPassword Saga", () => {
  let requestResetPasswordGenerator;

  beforeEach(() => {
    requestResetPasswordGenerator = requestResetPassword(
      requestResetPasswordPayload
    );
    const callDescriptor = requestResetPasswordGenerator.next(
      requestResetPasswordPayload
    ).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it("should dispatch the requestResetPasswordActionSuccess if it requests the data successfully", () => {
    const email = "randomString";
    const putDescriptor = requestResetPasswordGenerator.next(email).value;
    expect(putDescriptor).toEqual(
      put(requestResetPasswordActionSuccess(email))
    );
  });

  it("should call the requestResetPasswordActionError if the response errors", () => {
    const err = { response: { data: { message: "Test error" } } };
    const putDescriptor = requestResetPasswordGenerator.throw(err).value;
    expect(putDescriptor).toEqual(
      put(requestResetPasswordActionError("Test error"))
    );
  });
});
