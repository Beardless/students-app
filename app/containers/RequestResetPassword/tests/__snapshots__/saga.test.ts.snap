// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`requestResetPassword Saga should call the requestResetPasswordActionError if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      Object {
        "data": "{\\"email\\":\\"randomString\\"}",
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "POST",
        "url": "/users/password/reset_request",
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`requestResetPassword Saga should dispatch the requestResetPasswordActionSuccess if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      Object {
        "data": "{\\"email\\":\\"randomString\\"}",
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "POST",
        "url": "/users/password/reset_request",
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
